Documentation/Evolution/StringProcessingAlgorithms.md

```diff
@@ -0,0 +1,402 @@
+2. Generic `Collection` equivalents of these algorithms in terms of subsequences
+3. `protocol CustomRegexComponent`, allowing libraries to vend types that can be intermixed as components of regexes
+
+This proposal is part of a larger [regex-powered string processing initiative](https://forums.swift.org/t/declarative-string-processing-overview/52459). Throughout the document, we will reference the still-in-progress [`RegexProtocol`, `Regex`](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/Evolution/StronglyTypedCaptures.md), and [result builder DSL](https://forums.swift.org/t/pitch-regular-expression-literals/52820), but these are in flux and not formally part of this proposal. Further discussion of regex specifics is out of scope of this proposal and better discussed in another thread (see [Pitch and Proposal Status](https://github.com/apple/swift-experimental-string-processing/issues/107) for links to relevant threads).
+
+## Motivation
+
+TODO: Comparison for Swift and Python string utilities
+
+
+## Proposed solution
+
+### String algorithm additions
+
+We add frequently used string processing algorithms to the standard library. We also introduce internal infrastructure that allows groups of `Collection` algorithms that perform the same operations on different types to share their implementation, leading to a more coherent set of public APIs. This allows us to more easily provide algorithms that work with `RegexProtocol` values, such as
+
+```swift
+extension BidirectionalCollection where SubSequence == Substring {
+    public func ranges<R: RegexProtocol>(of regex: R) -> some Collection<Range<Index>>
+}
+```
+
+### Use custom parsers in regex builders and `RegexProtocol` algorithms
+
+We want to extend string processing to types from outside the standard library, so that one can use custom parsers with the string processing algorithms and incorporate custom parsers in regex builders seamlessly. 
+
+Consider parsing an HTTP header to capture the date field as a `Date` type:
+
+```swift
+let header = """
+HTTP/1.1 301 Redirect
+Date: Wed, 16 Feb 2022 23:53:19 GMT
+Connection: close
+Location: https://www.apple.com/
+"""
+```
+
+You are likely going to match a substring that look like a date string (`16 Feb 2022`), and parse the substring as a `Date` with one of the date parsers in the Foundation framework:
+
+```swift
+let regex = Regex {
+    capture {
+        oneOrMore(.digit)
+        " "
+        oneOrMore(.word)
+        " "
+        oneOrMore(.digit)
+    }
+}
+
+let dateParser = Date.ParseStrategy(format: "\(day: .twoDigits) \(month: .abbreviated) \(year: .padded(4))"
+if let dateMatch = header.firstMatch(of: regex)?.0 {
+    let date = try? Date(dateMatch, strategy: dateParser)
+}
+```
+
+This works, but wouldn't it be much more approachable if you can directly use the date parser within the string match function?
```

Comment:
```suggestion
While this approach happens to work for this example, it is fragile and more onerous than just directly using Foundation's date parser directly within the regex.
```

I would recommend against rhetorical questions, especially "wouldn't it be better" style ones.

Reply:
Also, you can consider streamlining some of this as part of an extended motivation section.